src/deno_dir.rs

```diff
@@ -0,0 +1,121 @@
+use fs;
+use sha1;
+use std;
+use std::path::Path;
+use std::path::PathBuf;
+
+// Example: /Users/rld/.deno/
+static mut ROOT: Option<PathBuf> = None;
+static mut DEPS: Option<PathBuf> = None;
+static mut GEN: Option<PathBuf> = None;
+pub enum Dirname {
+  Root,
+  // In the Go code this was called SrcDir. Renaming to Deps.
+  // Example: /Users/rld/.deno/deps/
+  Deps,
+  // In the Go code this was called CacheDir. Renaming to GEN.
+  // Example: /Users/rld/.deno/gen/
+  Gen,
+}
+pub use self::Dirname::*;
+
+pub fn path(dirname: Dirname) -> &'static Path {
+  setup().expect("deno_dir setup failed.");
+  let pb_option = unsafe {
+    match dirname {
+      Root => &ROOT,
+      Deps => &DEPS,
+      Gen => &GEN,
+    }
+  };
+  match pb_option {
+    Some(ref x) => x.as_path(),
+    None => panic!(),
+  }
+}
+
+// https://github.com/ry/deno/blob/golang/deno_dir.go#L99-L111
+pub fn setup() -> std::io::Result<()> {
+  // Only setup once.
+  if unsafe { ROOT != None } {
+    return Ok(());
+  }
+  let home_dir = std::env::home_dir().expect("Could not get home directory.");
+  // TODO(ry) Handle alternate deno dirs specified as command-line flag.
+  // unsafe because of mutable statics.
+
+  unsafe {
+    ROOT = Some(home_dir.join(".deno"));
+    DEPS = Some(path(Root).join("deps"));
+    GEN = Some(path(Root).join("gen"));
+  }
+  fs::mkdir(path(Gen))?;
+  fs::mkdir(path(Deps))?;
+
+  debug!("root {}", path(Root).display());
+  debug!("gen {}", path(Gen).display());
+  debug!("deps {}", path(Deps).display());
+
+  Ok(())
+}
+
+// https://github.com/ry/deno/blob/golang/deno_dir.go#L32-L35
+pub fn cache_path(filename: &str, source_code: &str) -> PathBuf {
+  let cache_key = source_code_hash(filename, source_code);
```

Comment:
So if a file's dependencies change, there's no need to recompile?
(Just asking, I don't know)

Reply:
That's right. In the compiled code the requires will reference the original module specifier, not the specifier of the compiled file.

Reply:
@piscisaureus Yeah, I don't think that happens... but let's see if we find any examples as we flesh out the tests.